clang/lib/Sema/SemaType.cpp

```diff
@@ -5347,13 +5347,16 @@ static TypeSourceInfo *GetFullTypeForDeclarator(TypeProcessingState &state,
 
         case NestedNameSpecifier::TypeSpec:
         case NestedNameSpecifier::TypeSpecWithTemplate:
-          ClsType = QualType(NNS->getAsType(), 0);
+          const Type *NNSType = NNS->getAsType();
+          ClsType = QualType(NNSType, 0);
           // Note: if the NNS has a prefix and ClsType is a nondependent
-          // TemplateSpecializationType, then the NNS prefix is NOT included
-          // in ClsType; hence we wrap ClsType into an ElaboratedType.
-          // NOTE: in particular, no wrap occurs if ClsType already is an
-          // Elaborated, DependentName, or DependentTemplateSpecialization.
-          if (isa<TemplateSpecializationType>(NNS->getAsType()))
+          // TemplateSpecializationType or a RecordType, then the NNS prefix is
+          // NOT included in ClsType; hence we wrap ClsType into an
+          // ElaboratedType. NOTE: in particular, no wrap occurs if ClsType
+          // already is an Elaborated, DependentName, or
+          // DependentTemplateSpecialization.
+          if (isa<TemplateSpecializationType>(NNSType) ||
+              (NNSPrefix && isa<RecordType>(NNSType)))
```

Comment:
```suggestion
              isa<RecordType>(NNSType))
```
We use an ElaboratedType with null prefix to represent a type which was written without qualifiers, so it seems arbitrary to not add the elaboration in this case.

Also, can you at least add a FIXME / TODO explaining how a DependentTemplateSpecializationType needs to be handled here, as I explained in the other thread?

Reply:
It would also be cool to have an assertion here, that we are handling all cases.

I think the logic here would be something like:
```
if (isa<DependentTemplateSpecializationType>...)
  // Rebuild DependentTemplateSpecializationType, adding the Prefix.
else
  // either the dependent case (TemplateSpecializationType), or the non-dependent one (RecordType).
 // assert(isa<TemplateSpecializationType, RecordType>...)
```

Reply:
> We use an ElaboratedType with null prefix to represent a type which was written without qualifiers, so it seems arbitrary to not add the elaboration in this case.

Wrapping RecordTypes into ElaboratedTypes with a null prefix was causing a bunch of failures on the first draft of the patch.

For example, on this line:

https://github.com/llvm/llvm-project/blob/8cb44859cc31929521c09fc6a8add66d53db44de/clang/test/CXX/conv/conv.mem/p4.cpp#L22

the change has the effect of changing the way the derived type is printed from `test1::Derived` to just `Derived`, I guess because when the diagnostic formatter encounters an `ElaboratedType` it says "I know how the type was printed in the source, I will print it the same way in the diagnostic".

Should we accept changes like this to diagnostic output?

Reply:
Yes exactly, this test change is correct, because the type was written as 'Derived' and not 'test1::Derived', and the diagnostics are supposed to print the type as written. It would be inconsistent to print the type as written when it has any qualifier, but print it fully qualified when it does not have any.

We started doing this consistently since this patch: https://reviews.llvm.org/D112374

Reply:
> It would also be cool to have an assertion here, that we are handling all cases.
> 
> I think the logic here would be something like:
> 
> ```
> if (isa<DependentTemplateSpecializationType>...)
>   // Rebuild DependentTemplateSpecializationType, adding the Prefix.
> else
>   // either the dependent case (TemplateSpecializationType), or the non-dependent one (RecordType).
>  // assert(isa<TemplateSpecializationType, RecordType>...)
> ```

I tried this, but `SemaCXX/err_init_conversion_failed.cpp` trips the assertion; we get a `TemplateTypeParmType` here.

Reply:
I see, well this is one case I left out :) The assert is still viable and would be helpful, but it's not critical for merging this.

Reply:
> The assert is still viable and would be helpful

FWIW, I did try adding back the assert to see what other sorts of types we get in this codepath. I found the following:

 * `TemplateTypeParmType` (seen in SemaCXX/err_init_conversion_failed.cpp)
 * `TypedefType` (seen in CXX/drs/cwg3xx.cpp)
 * `DecltypeType` (seen in CXX/expr/expr_prim/expr_prim_general/p8-0x.cpp)
 * `InjectedClassNameType` (seen in CXX/temp/temp_decls/temp_variadic/p5.cpp)
 * `UsingType` (seen in Modules/using-decl.cpp)

It's not clear to me what sort of handling we'd want for these various types.